src/roll.c

```diff
@@ -0,0 +1,143 @@
+
+  if (!LOGICAL(adaptive)[0]) {                                            // validating n input for adaptive=FALSE
+    if (isNewList(k))
+      error("n must be integer, list is accepted for adaptive TRUE");
+    //TODO move R's n=as.integer(n) to C
+    if (!isInteger(k))                                                    // check that k is integer vector
+      error("n must be integer");
+    i = 0;                                                                // check that all window values non-negative
+    while (i < nk && INTEGER(k)[i] >= 0) i++;
+    if (i != nk)
+      error("n must be non-negative integer values (>= 0)");
+  } else if (LOGICAL(adaptive)[0]) {                                      // validating n input for adaptive=TRUE
+    if (isVectorAtomic(k)) {                                              // if not-list then wrap into list
+      kl = allocVector(VECSXP, 1);
+      SET_VECTOR_ELT(kl, 0, k);
+      nk = 1;
+    } else {
+      kl = k;
+    }
+    //TODO move R's n=as.integer(n) to C
+    i = 0;                                                                // check that every column is integer type
+    while (i < nk && isInteger(VECTOR_ELT(kl, i))) i++;
+    if (i != nk)
+      error("n must be list of integer vectors when adaptive TRUE");
+  }
+  
+  if (length(fill) != 1)
+    error("fill must be a vector of length 1");
+
+  if (!strcmp(CHAR(STRING_ELT(align, 0)), "right"))
+    salign = RIGHT;
+  else if (!strcmp(CHAR(STRING_ELT(align, 0)), "center"))
+    salign = CENTER;
+  else if (!strcmp(CHAR(STRING_ELT(align, 0)), "left"))
+    salign = LEFT;
+  else
+    error("Internal error: invalid align argument in rolling function, should have been caught before. please report to data.table issue tracker.");
+
+  if (!isLogical(narm) || length(narm)!=1 || LOGICAL(narm)[0]==NA_LOGICAL)
+    error("na.rm must be TRUE or FALSE");
+  
+  ans = PROTECT(allocVector(VECSXP, nk * nx)); protecti++;                // allocate list to keep results
+  
+  thisfill = PROTECT(coerceVector(fill, REALSXP)); protecti++;
+  
+  if (!LOGICAL(adaptive)[0]) {                                          // adaptive=FALSE
+    if (salign == RIGHT) {                                              // align right scenario
+      for (i=0; i<nx; i++) {                                            // loop over columns
+        this = AS_NUMERIC(VECTOR_ELT(x, i));                            // extract column/vector from data.table/list
+        xrows = length(this);                                           // for list input each vector can have different length
+        #pragma omp parallel num_threads(MIN(getDTthreads(), nk))
+        {
+          #pragma omp for schedule(dynamic)
+          for (j=0; j<nk; j++) {                                          // loop over multiple windows
+            SET_VECTOR_ELT(ans, i*nk+j, tmp=allocVector(REALSXP, xrows)); // allocate answer vector for this column-window
```

Comment:
Correct: SET_VECTOR_ELT is definitely not thread safe and should never be used inside parallel code. What you observe is expected and correct behaviour; i.e. it might seem fine on small data or even big data with several threads. But as you increase the complexity, it will start to break in random ways.

Reply:
SET_VECTOR_ELT doesn't just assign,  but it updates the reference count on the object being assigned, and reduces the reference count on the object no longer pointed to by that slot.  When a `gc()` comes along, it's easiest to see that problems can occur then, but they can occur before then too. In addition to knowing this and doing things correctly,  as we get more multi-threaded, the test suite surviving `torture()` is getting more important.  This is why pruning the large tests was important,  to have a chance of torture < 24hrs.

Reply:
solved by separate loop to allocate storage for answer, then in pragma loop only `VECTOR_ELT` is used.

Reply:
solved by separate loop to allocate storage for answer, then in pragma loop only `VECTOR_ELT` is used.